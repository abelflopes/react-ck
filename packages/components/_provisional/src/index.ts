export * from "./image";

export * from "./avatar";

export * from "./dropdown";

export * from "./dock";

export * from "./breadcrumbs";

export * from "./scrollable-container";

export * from "./position-engine";

export * from "./top-bar";

export * from "./snackbar";

export * from "./overlay-spinner";

export * from "./menu";

export * from "./tooltip";

export * from "./popover";

export * from "./chat";

export * from "./file-uploader";

export * from "./attachment";

export * from "./snippet";
