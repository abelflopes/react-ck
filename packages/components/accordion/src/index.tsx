import React from "react";
import { List, type ListProps } from "@react-ck/list";
import { Collapse } from "@react-ck/collapse";
import styles from "./styles/index.module.scss";

export interface AccordionItem {
  header: NonNullable<React.ReactNode>;
  children: NonNullable<React.ReactNode>;
}

export interface AccordionProps extends Omit<ListProps, "items"> {
  items: AccordionItem[];
}

/**
 * An accordion is a vertically stacked list of interactive items. Each item can be "collapsed" with only a summary visible or “expanded” to show the full content for that item.
 * @param props - {@link AccordionProps}
 * @returns a React element
 */

export const Accordion = ({
  items,
  ...otherProps
}: Readonly<AccordionProps>): React.ReactElement => (
  <List
    {...otherProps}
    items={items.map(({ header, children }) => (
      <Collapse key={String(JSON.stringify(header))} header={header}>
        <div className={styles.content}>{children}</div>
      </Collapse>
    ))}
  />
);
