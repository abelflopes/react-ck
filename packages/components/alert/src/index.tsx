import React from "react";
import classNames from "classnames";
import styles from "./styles/index.module.scss";
import { Text } from "@react-ck/text";

export interface AlertProps extends React.HTMLAttributes<HTMLDivElement> {
  heading?: string;
  /** Specifies the visual style of the alert  */
  skin?: "neutral" | "primary" | "negative" | "average" | "positive";
}

// TODO: add dismiss

/**
 * Alert is a short message that provides contextual feedback in a prominent way.
 * @param props - {@link AlertProps}
 * @returns a React element
 */

export const Alert = ({
  heading,
  skin = "neutral",
  children,
  className,
  ...otherProps
}: Readonly<AlertProps>): React.ReactElement => (
  <div {...otherProps} className={classNames(styles.root, styles[skin], className)}>
    {heading && (
      <Text type="h4" as="p" variation="bold" margin={false} className={styles.heading}>
        {heading}
      </Text>
    )}
    {children}
  </div>
);
