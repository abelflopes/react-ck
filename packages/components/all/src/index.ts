export * from "./text";

export * from "./divider";

export * from "./banner";

export * from "./card";

export * from "./container";

export * from "./chip";

export * from "./collapse";

export * from "./list";

export * from "./overlay";
