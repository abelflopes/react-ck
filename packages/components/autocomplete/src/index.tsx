import React, { useEffect, useId, useMemo } from "react";
import styles from "./styles/index.module.scss";
import classNames from "classnames";
import { useFormFieldContext, type FormFieldProps } from "@react-ck/form-field";
import { AutocompleteOption } from "./AutocompleteOption";

interface AutocompleteProps extends React.InputHTMLAttributes<HTMLInputElement> {
  skin?: FormFieldProps["skin"];
}

/**
 * Autocomplete is a text input that proactively suggests options based on the userâ€™s initial input. The suggested options appear alongside the input in a panel.
 * @param props - {@link AutocompleteProps}
 * @returns a React element
 */

const Autocomplete = ({
  skin,
  id,
  className,
  children,
  ...props
}: Readonly<AutocompleteProps>): React.ReactElement => {
  const listId = useId();
  const formFieldContext = useFormFieldContext();

  const computedSkin = useMemo(
    () => formFieldContext?.skin ?? skin ?? "default",
    [formFieldContext?.skin, skin],
  );

  const computedId = useMemo(() => formFieldContext?.id ?? id, [formFieldContext?.id, id]);

  // Validate usage inside form field
  useEffect(() => {
    // Is not inside form field, skip
    if (formFieldContext === undefined) return;

    // Is inside form field
    if (skin)
      throw new Error("When using autocomplete inside form field, define skin on the form field");
    else if (id)
      throw new Error("When using autocomplete inside form field, define id on the form field");
  }, [formFieldContext, id, skin]);

  return (
    <>
      <input
        {...props}
        id={computedId}
        list={listId}
        className={classNames(
          styles.root,
          formFieldContext === undefined && styles.standalone,
          className,
          styles[`skin_${computedSkin}`],
        )}
      />
      <datalist id={listId}>{children}</datalist>
    </>
  );
};

Autocomplete.Option = AutocompleteOption;

export { Autocomplete, type AutocompleteProps };
