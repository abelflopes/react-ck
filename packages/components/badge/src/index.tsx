import classNames from "classnames";
import * as styles from "./styles/index.module.scss";
import React from "react";

export interface BadgeProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Specifies the visual style of the badge  */
  skin?: "neutral" | "primary" | "negative" | "average" | "positive";
}

/**
 * Badge is a compact label that appears beside a primary interface area which is used to represent status or metadata for that area..
 * @param props - {@link BadgeProps}
 * @returns a React element
 */

export const Badge = ({
  skin = "neutral",
  className,
  ...otherProps
}: Readonly<BadgeProps>): React.ReactElement => (
  <span {...otherProps} className={classNames(styles.root, styles[skin], className)} />
);
