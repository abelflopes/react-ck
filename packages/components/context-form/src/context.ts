import { createContext, useContext } from "react";

export interface ContextFormContextProps {
  setValue: (name: string, value: unknown) => void;
}

export const ContextFormContext = createContext<ContextFormContextProps>({
  setValue: () => {
    throw new Error("Context not initialized");
  },
});

export const useContextFormContext = (): ContextFormContextProps => useContext(ContextFormContext);
