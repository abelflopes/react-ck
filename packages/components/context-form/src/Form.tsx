import React, { useCallback, useMemo, useRef } from "react";
import { ContextFormContext, type ContextFormContextProps } from "./context";
import { Input } from "./fields/Input";
import { Select } from "@react-ck/select";
import { Textarea } from "@react-ck/textarea";

type Values = Record<string, unknown>;

interface ContextFormProps extends Omit<React.HTMLAttributes<HTMLFormElement>, "onSubmit"> {
  children: React.ReactNode;
  onValuesChange?: (values: Values) => void;
  onSubmit?: (props: { values: Values; e: React.FormEvent<HTMLFormElement> }) => void;
}

const ContextForm = ({
  children,
  onValuesChange,
  onSubmit,
  ...otherProps
}: Readonly<ContextFormProps>): React.ReactElement => {
  const values = useRef<Values>({});

  const setValue = useCallback<ContextFormContextProps["setValue"]>(
    (name, value) => {
      values.current = {
        ...values.current,
        [name]: value,
      };
      onValuesChange?.(values.current);
    },
    [onValuesChange],
  );

  const contextFormContextValue = useMemo<ContextFormContextProps>(
    () => ({
      setValue,
    }),
    [setValue],
  );

  return (
    <ContextFormContext.Provider value={contextFormContextValue}>
      <form
        {...otherProps}
        onSubmit={(e) => {
          onSubmit?.({ values: values.current, e });
        }}>
        {children}
      </form>
    </ContextFormContext.Provider>
  );
};

ContextForm.Input = Input;
ContextForm.Select = Select;
ContextForm.Textarea = Textarea;

export { ContextForm, type ContextFormProps };
