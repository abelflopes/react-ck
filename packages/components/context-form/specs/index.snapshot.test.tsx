import React from "react";
import { ContextForm } from "../src/index";
import renderer from "react-test-renderer";

describe("snapshot Context Form", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <ContextForm>
          <ContextForm.Input
            name="username"
            label="Username"
            placeholder="JohnDoe"
            description="Add a respectful alias name"
          />

          <ContextForm.Input
            name="email"
            type="email"
            label="Email"
            placeholder="john-doe@company.com"
            defaultValue="aaaaaa.s"
            description="Add your email address"
            skin="negative"
            validationMessage="Invalid email"
          />

          <ContextForm.Textarea
            name="address"
            label="Address"
            placeholder="St John's st. 12"
            description="Add your residential address"
          />
        </ContextForm>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
