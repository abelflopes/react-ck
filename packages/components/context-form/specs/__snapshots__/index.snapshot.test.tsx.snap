// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Context Form renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="root default"
  >
    <label
      className="root variation_small skin_bold margin_none label"
      htmlFor="aaa"
    >
      Username
    </label>
    <div
      className="content"
    >
      <input
        className="root"
        name="username"
        onChange={[Function]}
        placeholder="JohnDoe"
      />
    </div>
    <p
      className="root variation_small skin_default margin_none description"
    >
      Add a respectful alias name
    </p>
  </div>
  <div
    className="root negative"
  >
    <label
      className="root variation_small skin_bold margin_none label"
      htmlFor="aaa"
    >
      Email
    </label>
    <div
      className="content"
    >
      <input
        className="root"
        defaultValue="aaaaaa.s"
        name="email"
        onChange={[Function]}
        placeholder="john-doe@company.com"
        type="email"
      />
    </div>
    <p
      className="root variation_small skin_default margin_none description"
    >
      Add your email address
    </p>
    <p
      className="root variation_small skin_default margin_none validation_message"
    >
      Invalid email
    </p>
  </div>
  <div
    className="root default"
  >
    <label
      className="root variation_small skin_bold margin_none label"
      htmlFor="aaa"
    >
      Address
    </label>
    <div
      className="content"
    >
      <textarea
        className="root"
        name="address"
        placeholder="St John's st. 12"
      />
    </div>
    <p
      className="root variation_small skin_default margin_none description"
    >
      Add your residential address
    </p>
  </div>
</form>
`;
